Config

palako opened the pull request:
I've refactored the constants so that the view can be fully configurable.

I'm very new to Rust, I'm literally using your library in a project with the only purpose of learning the language, so please fell free to butcher my code for everything that is not done the Rust way. 

Maybe special attention to two of the constants that I've moved from &str to String, which I don't like, but I don't know how to go around the fact that to have the &str in the enum I had to add and propagate lifetimes everywhere, which I like even less, so the refactor to String is the less of the two evils as far as my limited knowledge goes at the moment.

Commit:
Refactor constants into configurable values

hellow554 commented on Cargo.toml:
```diff
@@ -9,11 +9,11 @@ keywords = ["cursive", "hexview"]
 [dependencies]
-cursive = { version = "0.14", default-features = false }
+cursive = { version = "0.15", default-features = false }
 itertools = "0.8"
 
 [dev-dependencies]
-cursive = "0.14"
+cursive = "0.15"
```
This is already done in your other PR, please rebase your commits!

hellow554 commented on examples/default.rs:
```diff
@@ -2,11 +2,10 @@ extern crate cursive;
 
 fn main() {
-    let mut cur = Cursive::default();
+    let mut cur = cursive::default();
```
Same here!

palako replied:
I needed this while working on these changes and it wasn't yet merged to your master, so I don't know how to do it differently or why it ended up in the pull request even though I created two different branches for it. I'll wait till you let me know how you want everything else, I'm not confident I won't make a mess if I rebase.

hellow554 replied:
It ended there, because it was in your branch ;) to get rid of it, you need to rebase your changes and delete that commit. Afterwards force push the changes. More info about rebasing can be found here: https://git-scm.com/book/en/v2/Git-Branching-Rebasing

palako replied:
Thanks for the link. I'm quite experienced with git and rebasing, it's just the whole github and pull requests thing that's very alien to me. I guess it's a matter of getting used to it and finding the few tricks that make it work. For example when you merged the other pull request, I updated my master with your master and now it says that mine is both one ahead and one behind!! go figure. I did find (this link)[https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request] which seems to go into the rebasing for pull requests. Wish me luck!

hellow554 commented on examples/hexdump.rs:
```diff
@@ -22,7 +21,7 @@ fn main() {
     let path = Path::new(&arg);
 
-    let mut cur = Cursive::default();
+    let mut cur = cursive::default();
```
Also this

hellow554 commented on examples/config.rs:
```diff
@@ -0,0 +1,35 @@
+    let explanation = TextView::new("Use the keys + - ↑ ↓ ← → 0-9 a-f for the HexView.\nUse q to exit.");
+    let data = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".as_bytes();
+    let view1 = HexView::new_from_iter(data).display_state(DisplayState::Editable);
+    let mut config = HexViewConfig {
+        hex_ascii_separator : " - ".to_owned(),
+        ..Default::default()
+    };
+
+    config.bytes_per_line = 8;
+    config.bytes_per_group = 4;
+    config.chars_per_spacing = 5;
```
any reason why those aren't merged? Let's merge them, looks better (and people who use rust should know how to mutate a struct with pub fields ;) )

palako replied:
The point I wanted to make was actually that you can use Default instead of having to change all, but as you said, maybe people just knows that stuff. I'm new to Rust so I thought it was useful to exemplify it like that. I've changed it now in the example file as part of [the change to have a non chainable version](https://github.com/palako/cursive_hexview/commit/1539f7e76ad27b4c684b7fe2677bb2c16ad99b08):

hellow554 commented on src/lib.rs:
```diff
@@ -140,9 +169,16 @@ impl HexView {
     }
 
+    ///Change configuration values for spacers and separators
+    pub fn config(mut self, config: HexViewConfig) -> Self {
+        self.config = config;
+        self
+    }
```
I also would like to see a `set_config` function which accepts a `&mut self` so modifying while not chaining things is easier.

palako replied:
Let me know if [this](https://github.com/palako/cursive_hexview/commit/1539f7e76ad27b4c684b7fe2677bb2c16ad99b08) is good.

hellow554 commented on src/lib.rs:
```diff
@@ -140,9 +169,16 @@ impl HexView {
         }
     }
 
+    ///Change configuration values for spacers and separators
```
I would like to see a small example, see the other comments of this struct

palako replied:
I'm not sure how to add value here without doing the same I did on the examples/config.rs file.

hellow554 replied:
Something like (please remove the space in the tripe backtick (```)

```
This function allows the customization of the `HexView` output.

For explanation of every possible option see the `HexViewConfig` struct.

#Examples

`` `
# use cursive_hexview::{DisplayState,HexView};
let view = HexView::new().config(HexViewConfig {
    bytes_per_line: 8,
    ..Default::default()
});
`` `

```

palako replied:
let me know if [this](https://github.com/palako/cursive_hexview/commit/935d10a3f180998a57bf694a8ba7c156a979ae4d) is good

hellow554 commented on src/lib.rs:
```diff
@@ -270,18 +306,12 @@ enum Field {
 
 /// calcs the position in a line with spacing
-fn get_cursor_offset(vec: Vec2) -> Vec2 {
+fn get_cursor_offset(vec: Vec2, config: HexViewConfig) -> Vec2 {
```
why is this accepted as value instead of reference? We don't need a copy of it, just pass a `&HexViewConfig` and you can omit a lot of `clone`s in the code

palako replied:
Let me know if [this](https://github.com/palako/cursive_hexview/commit/c31ca2e847910ce810c2c4b71ccbb5d274f7945a) is good

hellow554 commented on src/lib.rs:
```diff
@@ -332,17 +362,17 @@ impl HexView {
     /// and cursor (9, 0) will result in (11, 0) because of the 1+1 spacing after the fourth and eighth char
     fn get_cursor_offset(&self) -> Vec2 {
-        self.cursor + get_cursor_offset(self.cursor)
+        self.cursor + get_cursor_offset(self.cursor, self.config.clone())
```
here for example: this clone is unnecessary

palako replied:
yes, removed as part of the previous change.

hellow554 commented on src/lib.rs:
```diff
@@ -417,40 +447,41 @@ impl HexView {
+                    for &b in c {
+                        write!(&mut s, "{:02X}", b).expect("Unable to write hex values");
                     }
+                    format!("{}", s)
```
this can be simplified to `s.to_string()`

palako replied:
Let me know if [this](https://github.com/palako/cursive_hexview/commit/b745a36a1b20b9b35c2a643615c00f6884b0ce1c) is good

hellow554 commented:
> Maybe special attention to two of the constants that I've moved from &str to String, which I don't like

Yeah, I also don't like using a `String` here. I would probably use a `&'static str`.

If, in the future, people need a `String` we can later move to that or use a different approach (e.g. using `Cow`). But for now I would switch to `&'static str`

hellow554 commented on src/lib.rs:
```diff
@@ -59,6 +60,33 @@ pub enum DisplayState {
     Editable,
 }
 
+/// Configure spacing and separators for the view
```
I would like to see this commented extended. Please take a look at https://github.com/hellow554/cursive_hexview/blob/4168369ea2ff121520fa1a2b789ba2c0bdef4d40/src/lib.rs#L45-L60 for what I mean

palako replied:
Can you give me any direction or suggestion? I thought with the one-liners for each of the individual fields and the descriptive field names would be clear.

hellow554 replied:
```
Controls the visual output of the `HexView` struct.

There are various options which can be altered. For a detailed description of them, please see the fields below.
For the changes to apply, you need to use [`set_config`].

[`set_config`]: struct.HexView.html#method.set_config
```

hellow554 commented on src/lib.rs:
```diff
@@ -417,40 +447,41 @@ impl HexView {
     /// draws the hex fields between the addr and ascii representation
     fn draw_hex(&self, printer: &Printer) {
-        for (i, c) in self.data.chunks(U8S_PER_LINE).enumerate() {
+        let spacer = &format!("{:width$}", "", width=self.config.chars_per_spacing);
```
Please use https://doc.rust-lang.org/std/primitive.str.html#method.repeat instead, e.g.

`" ".repeat(self.config.chars_per_spacing)`

palako replied:
Let me know if [this](https://github.com/palako/cursive_hexview/commit/5b7ddb3fc73daa4ee163bb7cd5fe0caac2efb29a) is good

hellow554 replied:
I would probably put the `&` when using `spacer` and not while declaring it, but yes (:

hellow554 commented:
I think I got it now. Please rebase your changes (e.g. omit the 0.15 update commit), make the changes I requested and force push the changes. If you need help with that please message me and I will guide you through it.

One last request: Please use `cargo fmt` to format the code according to "my" guidelines. Thanks!

palako commented:
Hey Hellow554,
Thanks a lot for all of the comments. I've done most of the changes and pushed them to the branch in my fork so that you can have a look since I'm not sure if I'll make a whole mess of this pull request by pushing all if it. I've done things in different commits, but not sure if that's the best. I can always squash stuff I guess. I might need your input to a few things so that it all ends up looking the way you'd like it.
I haven't rebased the 0.15 change for the same reason, not sure what the best way to get all of this the one you want it is.
I'll add comments to your stuff above pointing to the individual commits, hoping it makes it easier for you to review.

Thanks!

palako commented:
> > Maybe special attention to two of the constants that I've moved from &str to String, which I don't like
> 
> Yeah, I also don't like using a `String` here. I would probably use a `&'static str`.
> 
> If, in the future, people need a `String` we can later move to that or use a different approach (e.g. using `Cow`). But for now I would switch to `&'static str`

Changed [here](https://github.com/palako/cursive_hexview/commit/4e7c8a3a8f2bde4c28430ffa18d7f265cead26f4). I didn't know that the solution to not propagate the lifetime was as easy as to declare it with a static lifetime. Cheers for that!!

palako commented:
I just realised that all the commits I pushed to my branch have been added to this pull request. Sorry, I'm not experienced with this!

hellow554 commented on src/lib.rs:
````diff
@@ -140,9 +169,46 @@ impl HexView {
+        self.config = config;
+    }
+
+    /// This function allows the customization of the `HexView` output. Chainable variant
+    ///
+    /// For options and explanation of every possible option, see the `HexViewConfig` struct.
+    ///
+    /// #Examples
+    ///
+    /// ```
+    /// # use cursive_hexview::{HexView, HexViewConfig};
+    /// let view = HexView::new().config(HexViewConfig {
+    ///     bytes_per_line: 8,
+    ///     ..Default::default()
+    /// });
+    ///```
````
Please do not repeat the doc exactly here, instead do something like 

https://github.com/hellow554/cursive_hexview/blob/4168369ea2ff121520fa1a2b789ba2c0bdef4d40/src/lib.rs#L174-L177

hellow554 commented on src/lib.rs:
````diff
@@ -140,9 +169,46 @@ impl HexView {
+    ///     ..Default::default()
+    /// });
+    ///```
+    pub fn config(mut self, config: HexViewConfig) -> Self {
+        self.config = config;
+        self
+    }
````
Please look at this example and adapt yours accordingly

https://github.com/hellow554/cursive_hexview/blob/4168369ea2ff121520fa1a2b789ba2c0bdef4d40/src/lib.rs#L175-L177

hellow554 commented:
> I just realised that all the commits I pushed to my branch have been added to this pull request.

this is exactly what it is supposed to be ;) So there is no need to always link to the commit you did something. 

> Sorry, I'm not experienced with this!

So the best thing you can do is contributing to something to learn it :) Thanks for all your work! I very much appreciate it.

hellow554 commented on src/lib.rs:
```diff
@@ -59,6 +60,33 @@ pub enum DisplayState {
+#[derive(Debug, Clone)]
+pub struct HexViewConfig {
+    /// Number of bytes per line
+    pub bytes_per_line: usize,
```
Controls how many bytes per line will be displayed. 

Question for you: What does happen on 0?

hellow554 commented on src/lib.rs:
```diff
@@ -59,6 +60,33 @@ pub enum DisplayState {
+/// Configure spacing and separators for the view
+#[derive(Debug, Clone)]
+pub struct HexViewConfig {
+    /// Number of bytes per line
```
Controls how many bytes per line will be displayed. 

    Default is 16

Question for you: What does happen on 0?

palako replied:
The same as if bytes_per_line is lower than bytes_per_group, curses doesn't even start. Do you think we should document it or do something else, like adding a check somewhere? Maybe even if just to add a more helpful error message.

hellow554 replied:
A panic is probably fine, when trying to display the view. If we want to make an error able way, we could add a `validate` method on the `HexViewConfig` struct and check that on set_config or on the view method itself.

hellow554 commented on src/lib.rs:
```diff
@@ -59,6 +60,33 @@ pub enum DisplayState {
+pub struct HexViewConfig {
+    /// Number of bytes per line
+    pub bytes_per_line: usize,
+    /// Number of bytes per group
```
Controls the number of bytes per group, before inserting a ` ` (0x20)

    Default is 1

TODO: Would it be feasible to add also a option for a `byte_group_separator` analogous to `addr_hex_seperator` etc?

palako replied:
I don't see why not, seems easy enough to replace the "".repeat for a variable. I don't see it very useful, but if you think it is, it costs nothing to add, I'm happy to do it. Probably in a different pull request I would say, to separate this one which was porting something fixed to be configurable, and the other one will be expanding the functionality to add that other option. But again, happy with what you prefer.

hellow554 replied:
Do it in this PR if that's okay for you. If you don't want to do it, I'm fine with it either, then I'll probably do it, but that's totaly fine.

hellow554 commented on src/lib.rs:
```diff
@@ -59,6 +60,33 @@ pub enum DisplayState {
+    /// Number of bytes per line
+    pub bytes_per_line: usize,
+    /// Number of bytes per group
+    pub bytes_per_group: usize,
+    /// Separation between groups
```
If we go for the `byte_group_separator` field, then this one will be obsolete, what do you think?

palako replied:
mmmm, not sure, I was thinking that the proposal for group separator would be same as we now have the separator between hex and ascii, which is a fixed value of a fixed length. In this case, if we do this, then if someone wants four spaces, they would need to set the group separator as "    ". So it is doable I guess, both ways would work. I like it how it  is right now better. And actually, even if adding the group separator, if we don't remove this one, then it's more flexible, even if it adds two ways of accomplishing the same thing:

```
1- group separator: " ", separation: 1 --> ab cd ef
2- group separator: " ", separation: 4 --> ab    cd    ef
3- group separator: "   ", separation: 1 --> ab    cd    ef
4- group separator: "<>", separation: 2 -> ab<><>cd<><>ef
5- group separator: "<><>", separation: 1 -> ab<><>cd<><>ef
```

If we leave it as it is, we can do 1 and 2
If we replace count for separator we can do 1, 3 and 5
If we keep both we can do all those 5.

2 and 3 are the same; 4 and 5 are the same, so both replacing or adding would allow all options. keeping the count just maintains that flexibility for the interface, which in my opinion is clearer since the " " is probably going to be the separator used 99.999999999999% of the times (the number of 9s is accurate :) ).

hellow554 replied:
Honstely, I would prefer having a new separator field 🤐 
Of course, most of the people would have a ` ` as separator, that's why it should be the `Default::default` case. Would you like to change the proposal to add a new field `byte_group_separator` and make it ` ` as default? The doc for it should be something like this:

```
Controls the separator between the hex groups in the data output.

Default is ` ` (0x20)
```

hellow554 commented on src/lib.rs:
```diff
@@ -59,6 +60,33 @@ pub enum DisplayState {
+    pub bytes_per_group: usize,
+    /// Separation between groups
+    pub chars_per_spacing: usize,
+    /// Separator between the address and data
```
```
Controls the separator between the address label and the hex output of the data.

Default is ` | `
```

hellow554 commented on src/lib.rs:
```diff
@@ -59,6 +60,33 @@ pub enum DisplayState {
+    pub chars_per_spacing: usize,
+    /// Separator between the address and data
+    pub addr_hex_separator: String,
+    /// Separator between the hex and ascii representations
```
```
Controls the separator between the hex output and the ASCII representation of the data.

Default is `: `
```

palako commented:
So to submit the changes you requested, what is the preferred way? Do i keep adding further commits to this branch and pull request? Or do I go back and amend the previous commits? The former it's definitely easier but I guess it leaves many commits in the history that should have never been there. The later is doable although a bit of a pain when there are, like is the case, several commits before and after the ones that need changing and re-commit-ing.

hellow554 commented:
> Do i keep adding further commits to this branch and pull request? 

Yes, for now, until we figured out what way we want to go. This makes it easier for me (and you hopefully) to track what's going on.

If we decide that this is good, I'd like to see a complete rebase of all commits into one or two commits, but for now let's keep going like you did before by pushing the fixes one by one.

palako commented:
There should be two new commits for you to review with what I believe is everything you pointed out. Please have a look and let me know. It would be pending the rebase to get rid of the 0.15 changes, which I'm waiting to do at the end when we are happy with everything else.

hellow554 commented on src/lib.rs:
```diff
@@ -60,18 +60,23 @@ pub enum DisplayState {
     pub addr_hex_separator: &'static str,
-    /// Separator between the hex and ascii representations
+    ///Controls the separator between the hex output and the ASCII representation of the data. Default is ` | ` (3 chars long)
     pub hex_ascii_separator: &'static str,
```
I did the newlines in my [previous proposal](https://github.com/hellow554/cursive_hexview/pull/7#discussion_r551460615) on purpose. The first line is always a short description on what it does, then a newline and then a detailed description. See https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html (and search for "first").

Use the following as an example to modify the doc of **all** fields

```
Controls the number of bytes per line.

It needs to be greater than 0 and equal or higher than the bytes_per group value.
Default is 16.
```

You can run `cargo doc --open` localy to see what it would look like in the end. It helps me to get a better understanding on what is linked and what not.

hellow554 replied:
@palako please see this comment. Please note the empty line after the first line!

hellow554 commented on src/lib.rs:
```diff
@@ -60,18 +60,23 @@ pub enum DisplayState {
 #[derive(Debug, Clone, Copy)]
 pub struct HexViewConfig {
-    /// Number of bytes per line
+    /// Controls the number of bytes per line. It needs to be greater than 0 and equal or higher than the bytes_per group value. Default is 16
```
Please write \`bytes_per_group\` so it will be linked automatically by rustdoc

Also a `.` is missing after 16

hellow554 commented on src/lib.rs:
```diff
@@ -60,18 +60,23 @@ pub enum DisplayState {
+    /// Controls the number of bytes per line. It needs to be greater than 0 and equal or higher than the bytes_per group value. Default is 16
     pub bytes_per_line: usize,
-    /// Number of bytes per group
+    /// Controls the number of bytes per group. It needs to be greater than 0 equal or lower than bytes_per_line. Default is 1.
```
Please write \`bytes_per_line\` so it will be linked automatically by rustdoc

hellow554 commented on src/lib.rs:
```diff
@@ -60,18 +60,23 @@ pub enum DisplayState {
-    /// Separation between groups
-    pub chars_per_spacing: usize,
-    /// Separator between the address and data
+    /// Controls the separator between the hex groups in the data output. Default is ` ` (1 char long)
```
Default is ` ` (0x20).

hellow554 commented on src/lib.rs:
```diff
@@ -60,18 +60,23 @@ pub enum DisplayState {
-    /// Separator between the address and data
+    /// Controls the separator between the hex groups in the data output. Default is ` ` (1 char long)
+    pub byte_group_separator: &'static str,
+    /// Controls the separator between the address label and the hex output of the data. Default is ` : ` (3 chars long)
```
Please remove _"(3 chars long)"_. It serves no purpose at all

palako replied:
I added it because of the spaces at either side are not very obvious when reading the doc, I thought it wouldn't hurt to clarify that the separator included those spaces. Nevermind, I have removed this in the last commit.

hellow554 replied:
Yes indeed, that's why it is in between the backticks:

    `   `

you the whitespaces will be displayed correctly and you can hover over them if you like to see how many spaces there really are.

hellow554 commented on src/lib.rs:
```diff
@@ -60,18 +60,23 @@ pub enum DisplayState {
+    /// Controls the separator between the address label and the hex output of the data. Default is ` : ` (3 chars long)
     pub addr_hex_separator: &'static str,
-    /// Separator between the hex and ascii representations
+    ///Controls the separator between the hex output and the ASCII representation of the data. Default is ` | ` (3 chars long)
```
Please remove _"(3 chars long)"_. It serves no purpose at all

hellow554 commented:
Alright! I think we're almost good to go.

I'm somewhat sorry that I have so many nitpick for you to solve. I hope you understand and bear with me.

Very good job and as said: I very much appretiate the work you've done so far.

palako commented:
That's no problem at all! The moment a single developer uses this and because of a properly written documentation he doesn't have to go browse the source code, all of this will have been worth it by saving them the time. I've also learnt at least 3 or 4 things about Rust during the review of this seemingly small change, so thank you!

hellow554 commented on src/lib.rs:
```diff
@@ -68,15 +68,20 @@ pub enum DisplayState {
 #[derive(Debug, Clone, Copy)]
 pub struct HexViewConfig {
-    /// Controls the number of bytes per line. It needs to be greater than 0 and equal or higher than the bytes_per group value. Default is 16
+    /// Controls the number of bytes per line. It needs to be greater than 0 and equal or higher than the `bytes_per_group` value.
+    /// Default is 16.
```
Unfortunately this isn't still quiet right. Please look at other doc example in the code and/or use the following as an example of what I mean:

```
    /// Controls the number of bytes per line.
    ///
    /// It needs to be greater than 0 and equal or higher than the `bytes_per_group` value.
    /// Default is 16.
    pub bytes_per_line: usize, 
```

As said, if you're unsure, use `cargo doc --open` and see if it looks like the other examples.

hellow554 commented on src/lib.rs:
```diff
@@ -68,15 +68,20 @@ pub enum DisplayState {
+    /// Default is 16.
     pub bytes_per_line: usize,
-    /// Controls the number of bytes per group. It needs to be greater than 0 equal or lower than bytes_per_line. Default is 1.
+    /// Controls the number of bytes per group. It needs to be greater than 0 equal or lower than `bytes_per_line`.
+    /// Default is 1.
```
Same here. Please make a newline after the first dot as well as an **empty** line (like above)

hellow554 commented on src/lib.rs:
```diff
@@ -68,15 +68,20 @@ pub enum DisplayState {
+    /// Default is 1.
     pub bytes_per_group: usize,
-    /// Controls the separator between the hex groups in the data output. Default is ` ` (1 char long)
+    /// Controls the separator between the hex groups in the data output.
+    /// Default is ` ` (0x20)
```
Same here

hellow554 commented on src/lib.rs:
```diff
@@ -68,15 +68,20 @@ pub enum DisplayState {
+    /// Default is ` ` (0x20)
     pub byte_group_separator: &'static str,
-    /// Controls the separator between the address label and the hex output of the data. Default is ` : ` (3 chars long)
+    /// Controls the separator between the address label and the hex output of the data.
+    /// Default is ` : `
```
Same here

hellow554 commented on src/lib.rs:
```diff
@@ -68,15 +68,20 @@ pub enum DisplayState {
+    /// Default is ` : `
     pub addr_hex_separator: &'static str,
-    ///Controls the separator between the hex output and the ASCII representation of the data. Default is ` | ` (3 chars long)
+    ///Controls the separator between the hex output and the ASCII representation of the data.
+    /// Default is ` | `
```
Same here

hellow554 commented on src/lib.rs:
```diff
@@ -59,6 +60,48 @@ pub enum DisplayState {
+///[`set_config`]: struct.HexView.html#method.set_config
+#[derive(Debug, Clone, Copy)]
+pub struct HexViewConfig {
+    /// Controls the number of bytes per line. It needs to be greater than 0 and equal or higher than the `bytes_per_group` value.
+    ///
+    /// Default is 16.
```
@palako your PR still isn't right. I mean.. I literally wrote what you have to write in https://github.com/hellow554/cursive_hexview/pull/7#discussion_r552550060 and still you somehow didn't do it.

Am I unclear to you or is it something else? Please correct this one.

hellow554 commented on src/lib.rs:
```diff
@@ -59,6 +60,48 @@ pub enum DisplayState {
+    ///
+    /// Default is 16.
+    pub bytes_per_line: usize,
+    /// Controls the number of bytes per group. It needs to be greater than 0 equal or lower than `bytes_per_line`.
+    ///
+    /// Default is 1.
```
Same here! The first line should only contain one sentence.

```
Controls the number of bytes per group.

It needs to be greater than 0 equal or lower than `bytes_per_line`.
Default is 1.
```

Exactly like this please

hellow554 commented:
Finally 🎉 

This look good. One last thing is, that you should run `cargo fmt` on the code, but that's something you can combine with the steps below.

The final and last part is to rebase/squash your changes. 
Please remove the 0.15 commit entirely, rebase against master and squash your commits down to one/two commits, whatever you feel like is appropiate.
Then force push the changes to your config branch and we're good to go.

If you have question, don't hesitate to ask.

palako commented:
I've squashed all the commits into one, and in the process I removed the 0.15 update. I run git hist on my local branch to verify that this was the case. Then I rebased from my master, which was updated from your master, and force pushed.
I see there are still two commits, one of them the 0.15.
Can you advice what I need to do to get rid of it?

palako commented:
The hash of the 0.15 commit is different than before, so it's definitely got rid of the previous one, but as part of the rebase, it's taken the new one and still added it to the pull request. I suspect it might be related to how I pulled that change from your master, which was a merge pull request with the rebase option, and left my master saying that I am one commit ahead and one commit behind your master.

hellow554 commented:
Hey, the 0.15 commit is still there, you can use `drop` instead of pick when doing a rebase.

Also I want a different commit message for the second one. It looks weird, that all the previous commits are listed there. Write a small text or leave it, as you like. I don't care much about git commit messages (at least here and in this PR, because it's very self explanatory).
You can use `edit` to edit the commit message while rebasing

palako commented:
Yes, I did use drop for that commit. This is how my hist looked in my local config branch before: 

```
$ git hist
* b087e23 2021-01-07 | cargo fmt (HEAD -> config) [Jose Palazon]
* bee0c57 2021-01-07 | move sentences arounds (origin/config) [Jose Palazon]
* 54023cf 2021-01-06 | reformat doc for config [Jose Palazon]
* b4c293d 2021-01-06 | Updates to documentation of the HexViewConfig struct [Jose Palazon]
* cf977e2 2021-01-05 | Refactor set_config and config implementations and documentation [Jose Palazon]
* 16dbe08 2021-01-05 | Replace byte group separator count of spaces for a custom configurable field [Jose Palazon]
* 935d10a 2021-01-04 | Add documentation for config functions [Jose Palazon]
* 5b7ddb3 2021-01-03 | refactor format! to repeat (highlight) [Jose Palazon]
* 4e7c8a3 2021-01-03 | refactor Strings to refs [Jose Palazon]
* b745a36 2021-01-03 | refactor format! to to_string [Jose Palazon]
* c31ca2e 2021-01-03 | Cursor offset to receive reference instead of value) [Jose Palazon]
* 1539f7e 2021-01-03 | add a non chainable version for set config [Jose Palazon]
* 7ed15fe 2020-12-24 | adjust the hardcoded bytes per line when editing [Jose Palazon]
* 0617abd 2020-12-24 | Refactor constants into configurable values [Jose Palazon]
* a559ca1 2020-12-23 | update to cursive-0.15 (origin/cursive-0.15, cursive-0.15) [Jose Palazon]
*   7ff3bd9 2020-04-11 | Merge pull request #5 from jwagen/jwagen-readme-version-update [Marcel Hellwig]
```

This is how it looked after the rebase (drop one commit and squash all others into one):

```
$ git hist
* eaaf5f8 2020-12-24 | Refactor constants into configurable values (HEAD -> config) [Jose Palazon]
*   7ff3bd9 2020-04-11 | Merge pull request #5 from jwagen/jwagen-readme-version-update [Marcel Hellwig]
```

And this is how it looks after the rebase:

```
$ git rebase master
First, rewinding head to replay your work on top of it...
Applying: Refactor constants into configurable values
(base) 192:cursive_hexview palako$ git hist
* 19ed8aa 2020-12-24 | Refactor constants into configurable values (HEAD -> config) [Jose Palazon]
* 74e7dcc 2020-12-31 | update to cursive-0.15 (#6) (origin/master, origin/HEAD, master) [JP]
*   7ff3bd9 2020-04-11 | Merge pull request #5 from jwagen/jwagen-readme-version-update [Marcel Hellwig]
```

Note how the commit ids `a559ca1` and `74e7dcc` for the "update to cursive-0.15" commit are different from the original to the rebased. Also the author, were Jose Palazon is my local user and JP is my github user.

It all looks "good" from a git perspective, but the commit is still added to the pull request.

hellow554 commented:
Don't rebase to master then ;) Just put the one commit here and it will be fine. Should work.

hellow554 commented:
Thank you very much for your work and patience with me!

That was awesome work and I'm very happy that this is now merged. I'm doing some polishing this weekend and hopefully release a new version then.

Thanks again! ❤️

palako commented:
No problem at all, thanks for your help and I'm glad everything ended as you like it, literally to the last dot and blank line! 👍 

I do have another feature I have added that I didn't want to send yet because of how messy this one was. It allows you to set a range of data to highlight. The way I use it in my app is so that when I select components in other views, say scrolling items on a list, I can highlight the hexadecimal data corresponding to the selected item in the list. Would you be interested in this change and if so would you like me to send the pull request now so that it makes it into the new version you want to release?

hellow554 commented:
@palako please open a new issue (with that text) and maybe a screenshot or two, so I can image, what you mean, but yeah, why not.

TBH: I don't want that feature in the next release, because I'm overdue with a new release, but there's nothing wrong with releasing a new version because of a new feature.
It's not like you need x amount of commits to release a new one, just grouping them into logical units.
